modules/fluid_properties/src/userobjects/CaloricImperfectGas.C

```diff
@@ -0,0 +1,970 @@
+{
+  const Real T = T_from_v_e(v, e);
+  const Real p = p_from_v_e(v, e);
+  Real dk_dp, dk_dT;
+  k_from_p_T(p, T, k, dk_dp, dk_dT);
+  // only dk_de = dk/dT * dT/de = 1/cv * dk/dT is nonzero
+  // (dk/dv)_e = 0 because e only depends on T and e is constant
+  dk_dv = 0.0;
+  dk_de = dk_dT / cv_from_p_T(p, T);
+}
+
+Real
+CaloricImperfectGas::s_from_v_e(Real v, Real e) const
+{
+  const Real T = T_from_v_e(v, e);
+  const Real p = p_from_v_e(v, e);
+  Real gamma = gamma_from_v_e(v, e);
+  const Real n = std::pow(T, gamma) / std::pow(p, gamma - 1.0);
+  if (n <= 0.0)
+    return getNaN("Negative argument in the ln() function.");
+  return cv_from_p_T(p, T) * std::log(n);
+}
+
+void
+CaloricImperfectGas::s_from_v_e(Real v, Real e, Real & s, Real & ds_dv, Real & ds_de) const
+{
+  Real T, dT_dv, dT_de;
+  T_from_v_e(v, e, T, dT_dv, dT_de);
+
+  Real p, dp_dv, dp_de;
+  p_from_v_e(v, e, p, dp_dv, dp_de);
+
+  Real gamma = gamma_from_v_e(v, e);
+
+  const Real n = std::pow(T, gamma) / std::pow(p, gamma - 1.0);
+  if (n <= 0.0)
+  {
+    s = getNaN("Negative argument in the ln() function.");
+    ds_dv = getNaN();
+    ds_de = getNaN();
+  }
+  else
+  {
+    Real cv = cv_from_p_T(p, T);
+
+    s = cv * std::log(n);
+
+    const Real dn_dT = gamma * std::pow(T, gamma - 1.0) / std::pow(p, gamma - 1.0);
+    const Real dn_dp = std::pow(T, gamma) * (1.0 - gamma) * std::pow(p, -gamma);
+
+    const Real dn_dv = dn_dT * dT_dv + dn_dp * dp_dv;
+    const Real dn_de = dn_dT * dT_de + dn_dp * dp_de;
+
+    ds_dv = cv / n * dn_dv;
+    ds_de = cv / n * dn_de;
+  }
+}
+
+Real
+CaloricImperfectGas::s_from_p_T(Real p, Real T) const
+{
+  Real gamma = gamma_from_p_T(p, T);
+  const Real n = std::pow(T, gamma) / std::pow(p, gamma - 1.0);
+  if (n <= 0.0)
+    return getNaN("Negative argument in the ln() function.");
+  return cv_from_p_T(p, T) * std::log(n);
+}
+
+void
+CaloricImperfectGas::s_from_p_T(Real p, Real T, Real & s, Real & ds_dp, Real & ds_dT) const
+{
+  Real gamma = gamma_from_p_T(p, T);
+  const Real n = std::pow(T, gamma) / std::pow(p, gamma - 1.0);
+  if (n <= 0.0)
+  {
+    s = getNaN("Negative argument in the ln() function.");
+    ds_dp = getNaN();
+    ds_dT = getNaN();
+  }
+  else
+  {
+    Real cv = cv_from_p_T(p, T);
+    s = cv * std::log(n);
+
+    const Real dn_dT = gamma * std::pow(T, gamma - 1.0) / std::pow(p, gamma - 1.0);
+    const Real dn_dp = std::pow(T, gamma) * (1.0 - gamma) * std::pow(p, -gamma);
+
+    ds_dp = cv / n * dn_dp;
+    ds_dT = cv / n * dn_dT;
+  }
+}
+
+Real
+CaloricImperfectGas::s_from_h_p(Real h, Real p) const
+{
+  Real T = T_from_p_h(p, h);
+  Real cv = cv_from_p_T(p, T);
+  Real gamma = gamma_from_p_T(p, T);
+  const Real aux = p * std::pow(h / (gamma * cv), -gamma / (gamma - 1));
+  if (aux <= 0.0)
+    return getNaN("Non-positive argument in the ln() function.");
+  return -(gamma - 1) * cv * std::log(aux);
+}
+
+// not sure it's correct to just replace _cv by cv(T(h))
```

Comment:
It probably needs a derivation if we're not sure

Reply:
This gets a bit more involved I think. 
```
First law: dq = dh - v dp
Second law: ds = dq/T
Ideal gas law: v = Rs T / p
==> ds = dh / T - Rs / p dp
Integrate but realize that dh = cp dT with non-constant cp:
s(h, p) - s0 = integral_h0^h dh / T - Rs log (p/p0)
```

I think we are free to select `s0` so set `s0=0` and that is consistent with 
`s(h0, p0) - s0 = integral_h0^h0 dh / T - Rs log (p0/p0) =0 => s(h0,p0)=s0`

We can still select `h0` and `p0` and we can simply pick `h0=hmin=h(Tmin)` and `p0=1Pa`.
Then we have
```
s(h,p) = integral_hmin^h dh / T(h) - Rs log(p)
```
 
The integral can be tabulated in terms of `h`. But it's not an analytical formula. 
Where do we need `s_from_h_p` and can I simply not implement it?

Reply:
> Where do we need s_from_h_p and can I simply not implement it?

We had to add this interface because of a boundary condition formulation involving stagnation properties. I don't even see it used in THM anymore - I think perhaps a CG BC formulation needed it. I wouldn't bother to implement it unless you find you need it.